### DISCUSSION TITLE
Why is the completion not resolved in the promise ?

### DISCUSSION THREAD

**Discussion by mspronesti** [Atome-FE/llama-node#37] [Q&A] [answered]
Hello and thanks a lot for this project !
I was just wondering why you don't resolve the `data` obeject in the promise [here](https://github.com/Atome-FE/llama-node/blob/61b841d02c9a37b09c2e2e0fd14c15a50a3c3c56/src/llm/llama-rs.ts#L43). In fact, I'd find it very useful to assign the completion result to a variable, which is not so convenient with a callback (in my humble opinion).

**Comment by hlhr202** [comment]
Hi, @mspronesti 

The reason for using callback here is to stream tokens one by one so that you can display it in realtime(once the backend has inferred the next token) instead of waiting the accumulation of tokens all together which may cost a lot of time. Generally, users would like to see the results directly even they came out in sequence, but users will get frustrated if they wait the results for over several seconds(in some slow machine it get even worse). Thats why we choose to implement the streaming like this.

The promise here is to notify the caller that all the inferencing progress has just completed. it is not used for accumulating tokens in sentences.

**Reply by mspronesti** [reply]
If you append the tokens to a vector and resolve it [here](https://github.com/Atome-FE/llama-node/blob/main/src/llm/llama-cpp.ts#L63) - similarly to what you do for errors - the user wouldn't wait the accumulation, would he ? 
The logic would be the same, but instead of resolving true / rejecting a vector of errors, you'd resolve the completion array / reject a vector error

**Reply by hlhr202** [reply]
its quite simple to add a non-streaming api for this and i can provide this probably next few iterations. but eventually the time delay comes from the model evaluation where it consumes a lot of computation resources, not from nodejs side.

**Reply by mspronesti** [reply]
That would help, thanks :)
However I still miss what difference would it make in terms of delay to resolve the vector of tokens (in the `.End` branch of the switch) instead of the boolean value. Isn't that the exact same streaming approach as before ?

**Reply by hlhr202** [reply]
emm i think i misunderstand your question from beginning. what you were asking is to keep the streaming but return the token strings instead of boolean, right? alright it really make sense and i may switch to this implementation in the future.

**Reply by mspronesti** [reply]
Yep, that's it. Thanks for taking it into considerations :)

**Reply by hlhr202** [reply, marked as answer]
https://github.com/Atome-FE/llama-node/pull/39/files#diff-d50d95d4d2e1defa9457fce8ea6f9d0672c5da30e533e069a998a9ce0f542ef2R72
will introduce new promise resolve/reject type like this
